doc/developer/guide-changes.md

```diff
@@ -0,0 +1,790 @@
+on a PR, however, the latency expectations for reviewing those changes are
+stricter. Prioritize reviewing those changes as soon as you are not in a flow
+state. Since you are already familiar with the PR, and typically the changes
+requested are small, reviewing the changes will typically very quick, in the
+5-15 minute range.
+
+Consider: if you're able to respond to a PR within an hour, that PR can go
+through three rounds of changes in an afternoon and land before EOD, freeing the
+author up to move on to something else the next day. If you only respond to a PR
+once a day, it will take the better part of a week to land the PR.
+
+## Software engineering principles
+
+This section details the software engineering principles we follow.
+Upholding these principles is a responsibility shared between the PR author and
+reviewer. The PR author should strive to adhere to these principles before
+submitting a PR for review, as much as is possible, and the reviewer should
+serve as a check to ensure these principles are followed. In particular, the
+reviewer should seek to contextualize these principles for the subsystem that
+is being changed, as these putting these principles into practice means
+different things in different parts of the codebase.
+
+### Simplicity
+
+The most important property a code can have is simplicity, bar none. Simple code
+is easier to understand and less likely to have bugs.
+
+Some amount of complexity is inevitable. After all, Materialize provides value
+to its customers by absorbing complexity. But the goal is to strip away all
+*incidental* complexity, leaving only the *essential* commplexity of the problem
+domain.
+
+The challenge is often defining what "simple" means. Which of these code samples
+is simpler?
+
+```rust
+let widgets: Vec<_> = gadgets
+    .into_iter()
+    .map(|g| g.frobulate())
+    .collect();
+```
+```rust
+let mut widgets = vec![];
+for g in gadgets {
+    widgets.push(g.frobulate());
+}
+```
+
+There are many experienced programmers on both sides of this argument, and the
+debate rages on.
+
+In many cases, however, simplicity is clear cut. Don't write this:
+
+```rust
+fn is_cold(temp: i32) -> bool {
+    if temp < 62 {
+        true
+    } else {
+        false
+    }
+}
+```
+
+Write this:
+
+```rust
+fn is_cold(temp: i32) -> bool {
+    temp < 62
+}
+```
+
+Whenever you prepare a PR, ask yourself: is the code I've written as simple as
+it could possibly be?
+
+A good mindset to have is that getting the code to work is only half the battle.
+You are only just getting started when the tests pass. If it takes you half a
+day to get the tests passing, it is entirely reasonable—desirable, even—to
+spend the next half of the day making the code as simple as possible and adding
+documentation/explanations.
+
+While ultimately our jobs involve adding features to Materialize, we are
+collectively responsible for minimizing the complexity of the system while we do
+so.
```

Comment:
I would add to this that minimizing complexity is part of adding features, because it makes it easier to add future features -- they're not orthogonal.

Reply:
Done.